### ISSUE TITLE
validate_numericality_of has problems with BigDecimal in Ruby 2.4

### ISSUE THREAD

**Issue by zavan** [thoughtbot/shoulda-matchers#986]
```ruby
Failure/Error:
       is_expected.to validate_numericality_of(:latitude)
         .is_greater_than_or_equal_to(-90).is_less_than_or_equal_to(90).allow_nil
     
     ArgumentError:
       invalid value for BigDecimal(): "abcd"
```

This exception was not thrown when doing `BigDecimal('abcd')` in Ruby < 2.4.0.

Ruby 2.4.0p0
Rails 5.0.1

**Comment by obahareth** [comment]
I've tried digging through the code but got lost going through the chain of files but I'm almost sure the culprit is [this line](https://github.com/thoughtbot/shoulda-matchers/blob/d1e9a41fca001b6d89455aff49f1f03ac272ee08/spec/unit/shoulda/matchers/active_model/validate_inclusion_of_matcher_spec.rb#L112).

Also if you look at the behavior of BigDecimal in Ruby 2.3

```ruby
BigDecimal.new("abcd")
=> #<BigDecimal:7fda565ae4d8,'0.0',9(9)>
```

It gives back a value that doesn't make a lot of sense, but now in Ruby 2.4 BigDecimal is actually validating input so that method for checking doesn't work.

Ruby 2.4:
```ruby
BigDecimal.new("abcd")
ArgumentError: invalid value for BigDecimal(): "abcd"
```

I wanted to try getting in a pull request for this but I can't get through a bundle install (yajl-ruby fails to install, I think it needs to be updated to a version above 1.3, but I couldn't find it in the gemfile and couldn't tell what was relying on it)

**Comment by obahareth** [comment]
@mcmire In my case, I can run my specs just fine on Ruby 2.3 but not in Ruby 2.4. I'm also using a decimal on PostgreSQL just like in #976. I'll give `bundle install` another go and see how it works out.

**Comment by obahareth** [comment]
I can run `bundle install` just fine now but `bundle exec rake` fails due to `json 1.8.3`, which apparently doesn't support Ruby 2.4.0. According to [this issue](https://github.com/flori/json/issues/303), version 2 supports it.

**Comment by zavan** [comment]
@obahareth Adding `gem 'json', '~> 2.0'` to my Gemfile fixes this specific problem.

**Comment by iwz** [comment]
Related: https://github.com/thoughtbot/shoulda-matchers/pull/996

**Comment by TheTeaNerd** [comment]
Is there any progress on this issue? We're still blocked from moving to Ruby 2.4.0 without dropping the shoulda-matchers

**Comment by horaciob** [comment]
@TheTeaNerd  Which version of rails you are using?

I've had same problem but seems not to be a shoulda error, BigDecimal has changed and this patch 
 https://github.com/rails/rails/commit/3aa729c2121b1e5a813afda52b2435dfcf7d4a74 was added on rails 5.0 but not in 4.X version.

For more information
https://github.com/rails/rails/pull/27429

**Comment by mcmire** [comment]
Would someone mind testing the `master` branch and seeing if this is still an issue? We have tests for `float` and `decimal` columns (which use BigDecimal, right?) and they are passing:

https://github.com/thoughtbot/shoulda-matchers/blob/2bab9dca080a707c1cf4de7c1b8d03ddde4ce323/spec/unit/shoulda/matchers/active_model/validate_numericality_of_matcher_spec.rb#L169-L181

**Comment by TheTeaNerd** [comment]
Ruby 2.4.2, Rails 5.0.0, shoulda-matchers `master`, BigDecimal 1.3.2 and I get:
```
Variant validations should validate that :reorder_trigger looks like a number less than or equal to 2147483647
     Failure/Error: it { is_expected.to validate_numericality_of(:reorder_trigger).is_less_than_or_equal_to(2_147_483_647) }

     ArgumentError:
       invalid value for BigDecimal(): "abcd"
```

Same on Rails 5 branch.

**Comment by mcmire** [comment]
@TheTeaNerd Hmm... okay. Thank you. I'll take another look and see if I can replicate this.

**Comment by mcmire** [comment]
@TheTeaNerd Sorry for the delay on this. I'm still not able to replicate this bug. I tried creating a new Rails 5.0.0 app using Ruby 2.4.1 making a model with a `bigint` column and typing in the test you gave and the test passed.

Are you using Postgres or MySQL? Can you give me some kind of backtrace so I can verify where this is coming from? Alternatively, if you could fork [this repo](https://github.com/mcmire/shoulda-matchers-test-apps) and see if you can reproduce the issue that would be wonderful 🌟

**Comment by choosen** [comment]
I met this issue with postgres

**Comment by guialbuk** [comment]
I think this issue has been solved by #1070 .

**Comment by mcmire** [comment]
@guialbuk That's what I initially thought, but then the error says `invalid value for BigDecimal(): "abcd"` and I think the only way that can happen is through this bit of code:

https://github.com/thoughtbot/shoulda-matchers/blob/afa6a7b6661a0d4450d330922c2c930ea2f3ff68/lib/shoulda/matchers/active_model/validate_numericality_of_matcher.rb#L499-L506

When this submatcher gets run it will end up setting the attribute to `"asdf"` (even if the column is a numeric column). This in itself shouldn't cause any problems; if you set a numeric column to a string that string will get typecast to 0. (I tried this in the Rails console, but only using Postgres -- that's why I was wondering about which database was being used.)

**Comment by mcmire** [comment]
@choosen Do you think you could replicate this in a fresh Rails app? https://github.com/mcmire/shoulda-matchers-test-apps

**Comment by choosen** [comment]
I tested on 'my' app and the problem occurred only at rails `4.2.8` 
(upgrade rails to `4.2.9` or `4.2.10` resolved issue)

`should validate_numericality_of(:price)`
```
minitest 5.11.3
shoulda 3.5.0
shoulda-matchers 2.8.0
```
@mcmire

**Comment by guialbuk** [comment]
It seems this problem has been fixed in Rails 4.2.9 by that commit: rails/rails@3aa729c

**Comment by mcmire** [comment]
Cool. I'm going to go ahead and close this issue for now, since I'm not able to replicate this under Rails 5.x, and also since there is a fix for Rails 4.2.x (we're not going to support older versions of 4.2.x for this release since it will be end-of-lifed soon anyway once Rails 6 is released).

If anyone continues to see this let me know - make sure to list which version of Ruby, Rails, and shoulda-matchers you're using.

**Comment by dotanlang** [comment]
Hi,
I seem to have this problem under Rails 5.0.1
I'm during a process of upgrading Ruby to 2.5.3 
And I just upgraded should-matchers to 4.0.1
Also using postgres
